src/mscorlib/src/System/Runtime/Loader/AssemblyLoadContext.cs

```diff
@@ -49,38 +65,75 @@ public abstract class AssemblyLoadContext
         internal static extern void InternalStartProfile(string profile, IntPtr ptrNativeAssemblyLoadContext);
 #endif // FEATURE_MULTICOREJIT
 
-        protected AssemblyLoadContext()
+        static AssemblyLoadContext()
         {
-            // Initialize the ALC representing non-TPA LoadContext
-            InitializeLoadContext(false);
+            // We register the cleanup of all AssemblyLoadContext that have not been finalized in the AppDomain.ProcessExit
+            AppDomain.CurrentDomain.ProcessExit += OnProcessExit;
         }
 
-        internal AssemblyLoadContext(bool fRepresentsTPALoadContext)
+        protected AssemblyLoadContext() : this(false)
         {
-            // Initialize the ALC representing TPA LoadContext
-            InitializeLoadContext(fRepresentsTPALoadContext);
         }
-        
-        void InitializeLoadContext(bool fRepresentsTPALoadContext)
+
+        internal AssemblyLoadContext(bool fRepresentsTPALoadContext)
         {
-            // Initialize the VM side of AssemblyLoadContext if not already done.
-            GCHandle gchALC = GCHandle.Alloc(this);
-            IntPtr ptrALC = GCHandle.ToIntPtr(gchALC);
-            m_pNativeAssemblyLoadContext = InitializeAssemblyLoadContext(ptrALC, fRepresentsTPALoadContext);
+            // Add this instance to the list of alive ALC
+            lock (assemblyLoadContextAliveList)
+            {
+                if (isProcessExiting)
+                {
+                    // TODO: Add localized version
+                    throw new InvalidOperationException("Cannot create an AssemblyLoadContext when the process is exiting");
+                }
 
-            // Initialize event handlers to be null by default
-            Resolving = null;
-            Unloading = null;
+                // Initialize the VM side of AssemblyLoadContext if not already done.
+                isDefault = fRepresentsTPALoadContext;
+                var thisWeakHandle = GCHandle.Alloc(this, GCHandleType.Weak);
```

Comment:
I'm wondering where is the overhead exactly coming from? (the GC that has more work to do on a collectible assembly?).

Currently the unload is automatically done on the ALC finalizer if not called explicitly. So that would mean that if an ALC is not unloadable, we would throw an exception on the Unload method? Behind the scene, the only difference would be that the native Assemblies + AssemblyLoadContext + AssemblyLoaderAllocator would not me marked collectible I guess.